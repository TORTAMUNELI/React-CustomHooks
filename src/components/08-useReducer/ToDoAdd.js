import React from 'react'
import { useForm } from '../../hooks/useForm';

export const ToDoAdd = ({ handleAddToDo }) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) return;

        const newToDo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddToDo(newToDo);

        reset();
    }

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    className='form-control'
                    type='text'
                    name='description'
                    placeholder='Aprender ...'
                    autoComplete='off'
                    value={description}
                    onChange={handleInputChange}
                />
                <div className="d-grid gap-2 mt-3">
                    <button
                        type='submit'
                        className="btn btn-outline-primary"
                    >
                        Agregar
                    </button>
                </div>
            </form>
        </>
    )
}
