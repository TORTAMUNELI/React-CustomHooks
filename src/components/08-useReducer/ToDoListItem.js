import React from 'react'
import PropTypes from 'prop-types';

export const ToDoListItem = ({ index, toDo, handleDelete, handleToggle }) => {
    return (
        <li
            key={toDo.id}
            className='list-group-item'
        >
            <p
                className={`${toDo.done && 'complete'}`}
                onClick={() => handleToggle(toDo.id)}
            >{index + 1}. {toDo.desc}</p>
            <button
                className='btn btn-danger'
                onClick={() => handleDelete(toDo.id)}
            >
                Borrar
            </button>
        </li>
    )
}

ToDoListItem.propTypes = {
    index: PropTypes.number.isRequired,
    toDo: PropTypes.object.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}