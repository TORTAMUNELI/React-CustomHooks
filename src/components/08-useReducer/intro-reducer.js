const initialState = [{
    id: 1,
    toDo: 'Comprar Leche',
    done: false
}];

const toDoReducer = (state = initialState, action) => {

    if (action?.type === 'agregar') {
        return [...state, action.payload];
    }

    return state;
}

let toDos = toDoReducer();

const newToDo = {
    id: 2,
    toDo: 'Comprar Pan',
    done: false
};

const agregarToDoAction = {
    type: 'agregar',
    payload: newToDo
};

toDos = toDoReducer(toDos, agregarToDoAction);

console.log(toDos);