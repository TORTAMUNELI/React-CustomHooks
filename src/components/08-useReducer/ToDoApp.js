import React, { useReducer } from 'react'
import { toDoReducer } from './toDoReducer';
import { ToDoList } from './ToDoList';
import { ToDoAdd } from './ToDoAdd';
import { useEffect } from 'react/cjs/react.development';
import './styles.css';


const init = () => {
    return JSON.parse(localStorage.getItem('toDos')) || [];
}

export const ToDoApp = () => {

    const [toDos, dispatch] = useReducer(toDoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('toDos', JSON.stringify(toDos));
    }, [toDos])

    const handleDelete = (toDoId) => {
        const action = {
            type: 'delete',
            payload: toDoId
        }

        dispatch(action);
    }

    const handleToggle = (toDoId) => {
        dispatch({
            type: 'toggle',
            payload: toDoId
        });
    }

    const handleAddToDo = (newToDo) => {
        dispatch({
            type: 'add',
            payload: newToDo
        });
    }

    return (
        <>
            <h1>ToDoApp ({toDos.length})</h1>
            <hr />

            <div className='row'>
                <div className='col-7'>
                    <ToDoList
                        toDos={toDos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                </div>

                <div className='col-5'>
                    <ToDoAdd
                        handleAddToDo={handleAddToDo}
                    />
                </div>
            </div>
        </>
    )
}
