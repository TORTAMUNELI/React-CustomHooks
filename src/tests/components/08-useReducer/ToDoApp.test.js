import React from 'react'
import { shallow } from 'enzyme';
import { ToDoApp } from '../../../components/08-useReducer/ToDoApp'

describe('Pruebas en <ToDoApp />', () => {
    const wrapper = shallow(<ToDoApp />);

    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

});
