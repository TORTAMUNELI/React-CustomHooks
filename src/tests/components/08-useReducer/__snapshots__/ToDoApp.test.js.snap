// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ToDoApp /> Debe mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    ToDoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <ToDoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        toDos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <ToDoAdd
        handleAddToDo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
